File: OJ/leet1004/leet1004.py
# leet1004

from typing import List


class Solution:
    def longestOnes(self, A: List[int], K: int) -> int:
        # f(l, r) = # number of 0s in A[l, r]
        lft, rht = 0, 0
        f = 0
        ans = 0
        while rht < len(A):
            if A[rht] == 0:
                f += 1
                while f > K:
                    if A[lft] == 0: f -= 1
                    lft += 1
            rht += 1
            ans = max(ans, rht-lft)
        return ans




print(Solution().longestOnes(A = [1,1,1,0,0,0,1,1,1,1,0], K = 2))
print(Solution().longestOnes(A = [0,0,1,1,0,0,1,1,1,0,1,1,0,0,0,1,1,1,1], K = 3))

File: OJ/leet1004/leet1004.cpp
/*
leet1004


*/


#include <iostream>
#include <tuple>
#include <vector>
#include <queue>
#include <stack>
#include <string>
#include <unordered_set>
#include <unordered_map>
#include <algorithm>
#include <numeric>
#include <cassert>
#include <utility>
// #include "../utils/LeetCpp.utils.hpp"

using namespace std;

class Solution {
public:
    int longestOnes(const vector<int>& A, int K) {
        int lft = 0, n = A.size(), ans = 0;
        for (int rht = 0; rht < n; rht++)
        {
            if (A[rht] == 0) {
                if (K == 0) {
                    while (A[lft] == 1) lft++;
                    lft++;
                }
                else K--;
            }
            ans = max(ans, rht - lft + 1);
        }
        return ans;
    }
};

int main(int argc, char const *argv[])
{
    Solution sol;   
    cout << sol.longestOnes({1,1,1,0,0,0,1,1,1,1,0}, 2) << endl;
    cout << sol.longestOnes({0,0,1,1,0,0,1,1,1,0,1,1,0,0,0,1,1,1,1}, 3) << endl;
    return 0;
}
